lib/puppet/util/windows/registry.rb

```diff
@@ -30,51 +32,337 @@ def open(name, path, mode = KEY_READ | KEY64, &block)
+          end
+        end
+      end
+
+      [subkey, filetime]
+    end
+
+    def reg_enum_value(key, index, max_value_length = Win32::Registry::Constants::MAX_VALUE_LENGTH)
+      subkey, type, data = nil, nil, nil
+
+      FFI::MemoryPointer.new(:dword) do |subkey_length_ptr|
+        FFI::MemoryPointer.new(:wchar, max_value_length) do |subkey_ptr|
+          # RegEnumValueW cannot be called twice to properly size the buffer
+          subkey_length_ptr.write_dword(max_value_length)
+
+          result = RegEnumValueW(key.hkey, index,
+            subkey_ptr, subkey_length_ptr,
+            FFI::Pointer::NULL, FFI::Pointer::NULL,
+            FFI::Pointer::NULL, FFI::Pointer::NULL
+          )
+
+          break if result == ERROR_NO_MORE_ITEMS
+
+          if result != FFI::ERROR_SUCCESS
+            msg = "Failed to enumerate #{key.keyname} registry values at index #{index}"
+            raise Puppet::Util::Windows::Error.new(msg)
+          end
+
+          subkey_length = subkey_length_ptr.read_dword
+          subkey = subkey_ptr.read_wide_string(subkey_length, Encoding::UTF_8)
+
+          type, data = read(key, subkey_ptr)
+        end
+      end
+
+      [subkey, type, data]
+    end
+
+    def reg_query_info_key_max_lengths(key)
+      result = nil
+
+      FFI::MemoryPointer.new(:dword) do |max_subkey_name_length_ptr|
+        FFI::MemoryPointer.new(:dword) do |max_value_name_length_ptr|
+
+          status = RegQueryInfoKeyW(key.hkey,
+            FFI::MemoryPointer::NULL, FFI::MemoryPointer::NULL,
+            FFI::MemoryPointer::NULL, FFI::MemoryPointer::NULL,
+            max_subkey_name_length_ptr, FFI::MemoryPointer::NULL,
+            FFI::MemoryPointer::NULL, max_value_name_length_ptr,
+            FFI::MemoryPointer::NULL, FFI::MemoryPointer::NULL,
+            FFI::MemoryPointer::NULL
+          )
+
+          if status != FFI::ERROR_SUCCESS
+            msg = "Failed to query registry #{key.keyname} for sizes"
+            raise Puppet::Util::Windows::Error.new(msg)
+          end
+
+          result = [
+            # Unicode characters *not* including trailing NULL
+            max_subkey_name_length_ptr.read_dword + 1,
+            max_value_name_length_ptr.read_dword + 1,
+          ]
+        end
+      end
+
+      result
+    end
+
+    # Read a registry value named name and return array of
+    # [ type, data ].
+    # When name is nil, the `default' value is read.
+    # type is value type. (see Win32::Registry::Constants module)
+    # data is value data, its class is:
+    # :REG_SZ, REG_EXPAND_SZ
+    #    String
+    # :REG_MULTI_SZ
+    #    Array of String
+    # :REG_DWORD, REG_DWORD_BIG_ENDIAN, REG_QWORD
+    #    Integer
+    # :REG_BINARY
+    #    String (contains binary data)
+    #
+    # When rtype is specified, the value type must be included by
+    # rtype array, or TypeError is raised.
+    def read(key, name_ptr, *rtype)
```

Comment:
Yeah, it was already called by things higher in the call stack, so I don't think we needed to expose it.  Also note that I changed the Ruby implementation a little to optimized for taking an `FFI::MemoryPointer` instead of a Ruby `String`, so that we could avoid an extra few unnecessary allocations / round-tripping through Ruby's UTF-8 process.  That was another +1 for the `keep it private` argument.